### PR TITLE
Feature: support of asciinema cast files

### PR DESCRIPTION
**Pull request by tdaron**
I built this feature for fun and to see how eplbot was structured, but i still think it can be useful on epl server !

# How it works

When a user upload a .cast file recorded using [asciinema](https://asciinema.org/) , the bot will automatically convert it to a gif, and upoad it on discord, with a quote of user message, and it will delete original user message containing the cast. 

# Requirements

This feature requires you to have [agg](https://github.com/asciinema/agg) installed. 
I wanted to be sure the feature was interesting to you before putting effort of adding agg to  the Dockerfile

### CODE REVIEW

**Comment by Hokkaydo** [comment]
It looks amazing. I didn't know about asciinema.
But is there a lot of people using this? 
I've never seen anyone posting a .cast file

**Comment by tdaron** [comment]
> It looks amazing. I didn't know about asciinema. But is there a lot of people using this? I've never seen anyone posting a .cast file

Well, it's quite popular on github README's to demonstrate usage of a cli tool

but in epl i don't think so, but mainly because people don't know (yet) this tool i think.  

(and the .cast file is pretty useless by its own, it has to be converted to somewhat like a gif, which is done by agg. cast files are just text files containing terminal output diff at each timestamp)

**Comment by Hokkaydo** [comment]
> > It looks amazing. I didn't know about asciinema. But is there a lot of people using this? I've never seen anyone posting a .cast file
> 
> Well, it's quite popular on github README's to demonstrate usage of a cli tool
> 
> but in epl i don't think so, but mainly because people don't know (yet) this tool i think.
> 
> (and the .cast file is pretty useless by its own, it has to be converted to somewhat like a gif, which is done by agg. cast files are just text files containing terminal output diff at each timestamp)

Indeed, we may give a try

File: .gitignore
Comment (Hokkaydo) [inline .gitignore:1]: Rebase on dev as it has been already fixed by #49

File: Dockerfile
Comment (Hokkaydo) [inline Dockerfile:1]: Rebase on dev as it has been already fixed by #49

File: README.md
Comment (Hokkaydo) [inline README.md:1]: Rebase on dev as it has been already fixed by #49

File: src/main/java/com/github/hokkaydo/eplbot/module/asciinema/AsciinemaListener.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    }
+    private static final Random random = new Random();
+
+
+    @Override
+    public void onMessageReceived(@NotNull MessageReceivedEvent event) {
+        List<Message.Attachment> attachments = event.getMessage().getAttachments();
+        if (attachments.isEmpty()) return; // no attachments on the message!
+        for (Message.Attachment att: attachments) {
+            if (Objects.equals(att.getFileExtension(), "cast")) {
```
Comment (Hokkaydo) [inline src/main/java/com/github/hokkaydo/eplbot/module/asciinema/AsciinemaListener.java]: Use early return

File: src/main/java/com/github/hokkaydo/eplbot/module/asciinema/AsciinemaListener.java [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+                builder.command("agg", fileName, gifFileName);
+                try {
+                    builder.start().waitFor();
+                } catch (InterruptedException | IOException e) {
+                    assert (new File(fileName)).delete();
+                    throw new RuntimeException(e);
+                }
+                File gifFile = new File(gifFileName);
+                Member author = event.getMember();
+                assert author != null;
```
Comment (Hokkaydo) [inline src/main/java/com/github/hokkaydo/eplbot/module/asciinema/AsciinemaListener.java]: Do a proper check
```suggestion
                if (author == null) return;
```

File: src/main/java/com/github/hokkaydo/eplbot/module/asciinema/AsciinemaListener.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+import java.util.concurrent.ExecutionException;
+import java.util.function.Consumer;
+
+public class AsciinemaListener extends ListenerAdapter {
+
+    private final Long guildId;
+    AsciinemaListener(Long guildId) {
+        this.guildId = guildId;
+    }
+    private static final Random random = new Random();
```
Comment (Hokkaydo) [inline src/main/java/com/github/hokkaydo/eplbot/module/asciinema/AsciinemaListener.java]:
```suggestion
    private static final Random RANDOM = new Random();
```

**Review by Hokkaydo — CHANGES_REQUESTED** [review]
I don't like to have a lot of try-catch blocks. 
We should consider using the Stream API. However, I can't immediately find a nice way to do it. I need more reflection.
Aside from that, it looks great. I batched what I could but I can't batch the early return as there is a brace to remove at the end of the block

File: src/main/java/com/github/hokkaydo/eplbot/module/asciinema/AsciinemaListener.java (hunk 1 above)
Comment (tdaron) [reply src/main/java/com/github/hokkaydo/eplbot/module/asciinema/AsciinemaListener.java]: wdym by early return ? If i find one cast file i return from the function ?

File: .gitignore
Comment (tdaron) [reply .gitignore:1]: done 1

File: Dockerfile
Comment (tdaron) [reply Dockerfile:1]: done 2

File: README.md
Comment (tdaron) [reply README.md:1]: done 3

File: src/main/java/com/github/hokkaydo/eplbot/module/asciinema/AsciinemaListener.java (hunk 1 above)
Comment (Hokkaydo) [reply src/main/java/com/github/hokkaydo/eplbot/module/asciinema/AsciinemaListener.java]: Actually "early continue" rather
`if (!Object.equals(att.getFileExtension(), "cast")) continue;`
It helps keeping the code readable by removing one level of indentation

**Comment by tdaron** [comment]
early continue + comments: done

**Comment by Hokkaydo** [comment]
Could you consider adding "agg" in the docker image?

**Comment by tdaron** [comment]
done !

**Pull request closed without merging**